fw/access_log.c

```diff
@@ -0,0 +1,246 @@
+
+	n = tfw_str_to_cstr(s, p, limit);
+	if (n + 1 == limit && n >= 3) {
+		p[n - 3] = '.';
+		p[n - 2] = '.';
+		p[n - 1] = '.';
+	}
+	p += n;
+
+	return n;
+}
+
+
+void
+do_access_log(TfwHttpResp *resp)
+{
+	do_access_log_req(resp->req, resp->status, resp->content_length);
+}
+
+
+void
+do_access_log_req(TfwHttpReq *req, int status, unsigned long content_length)
+{
+	char *buf = this_cpu_ptr(access_log_buf);
+	char *p = buf, *end = p + ACCESS_LOG_BUF_SIZE;
+
+	if (!access_log_enabled)
+		return;
+
+#define CONCAT_TFW_STR(s, remaining, reserve)                      \
+	p += append_to_cstr_bounded(p, end, s, remaining, reserve)
+#define CONCAT_HDR(hdr_id, remaining, reserve) do {                \
+                TfwStr hdr = get_http_header_value(req->version,   \
+				req->h_tbl->tbl + hdr_id);         \
+                CONCAT_TFW_STR(&hdr, remaining, reserve);          \
+	} while (0)
+#define CONCAT_PRINTF(fmt, ...) do {                               \
+		p += snprintf(p, end - p, fmt, ##__VA_ARGS__);     \
+		if (p >= end) goto overflow;                       \
+	} while (0)
+#define CONCAT_STR(str) do {                                       \
+		if (p + sizeof(str) >= end) goto overflow;         \
+		memcpy(p, str, sizeof(str) - 1);                   \
+		p += sizeof(str) - 1;                              \
+	} while (0)
+
+	if (req)
+		TODO_LOG_CONN(req);
+	/* Resp->conn is NULL if invalid response had been received */
+	if (req->conn && req->conn->peer)
+		p = tfw_addr_fmt(&req->conn->peer->addr, TFW_NO_PORT, p);
+	else
+		CONCAT_STR("-");
+
+	CONCAT_STR(" \"");
+	if (req->vhost != NULL && !TFW_STR_EMPTY(&req->vhost->name))
+		CONCAT_TFW_STR(&req->vhost->name, 3, 40);
+	else
+		CONCAT_STR("-");
+#define CONCAT_CASE_STR(x, str) case x: CONCAT_STR(str); break
+
+#define CONCAT_CASE_STR_XX(x, str) CONCAT_CASE_STR(x, XX(str))
+	switch (req->method) {
+#define XX(str) "\" \"" str " "
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_COPY,       "COPY");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_DELETE,     "DELETE");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_GET,        "GET");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_HEAD,       "HEAD");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_LOCK,       "LOCK");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_MKCOL,      "MKCOL");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_MOVE,       "MOVE");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_OPTIONS,    "OPTIONS");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_PATCH,      "PATCH");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_POST,       "POST");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_PROPFIND,   "PROPFIND");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_PROPPATCH,  "PROPPATCH");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_PUT,        "PUT");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_TRACE,      "TRACE");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_UNLOCK,     "UNLOCK");
+	CONCAT_CASE_STR_XX(TFW_HTTP_METH_PURGE,      "PURGE");
+	default: CONCAT_STR("-");
+#undef XX
+	}
+	if (!TFW_STR_EMPTY(&req->uri_path))
+		CONCAT_TFW_STR(&req->uri_path, 3,
+			       8 + 2 + 10 + 1 + 20 + 2 + 3 + 1);
```

Comment:
Please comment the constants or use defines instead. Also for the same call with with the constant `40` above